include/zephyr/modules/module.h

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_MODULE_H
+#define ZEPHYR_MODULE_H
+
+#include <zephyr/sys/slist.h>
+#include <zephyr/modules/elf.h>
+#include <sys/types.h>
+
+/**
+ * @brief A module symbol type tag
+ */
+enum module_symbol_type {
+	MODULE_SYMBOL_FUNC,
+	MODULE_SYMBOL_VAR,
+	MODULE_SYMBOL_UNDEF_FUNC,
+};
+
+/**
+ * @brief A symbol (named memory address)
+ */
+struct module_symbol {
+	enum module_symbol_type tt;
+	char name[32];
+	void *addr;
+};
+
+/**
+ * @brief A symbol table
+ *
+ * An array of symbols
+ */
+struct module_symtable {
+	elf_word sym_cnt;
+	struct module_symbol *syms;
+};
+
+
+/**
+ * @brief Enum of module memory regions for lookup tables
+ */
+enum module_mem {
+	MOD_MEM_TEXT,
+	MOD_MEM_DATA,
+	MOD_MEM_RODATA,
+	MOD_MEM_BSS,
+
+	MOD_MEM_COUNT,
+};
+
+/**
+ * @brief Enum of sections for lookup tables
+ */
+enum module_section {
+	MOD_SECT_TEXT,
+	MOD_SECT_DATA,
+	MOD_SECT_RODATA,
+	MOD_SECT_BSS,
+
+	MOD_SECT_REL_TEXT,
+	MOD_SECT_REL_DATA,
+	MOD_SECT_REL_RODATA,
+	MOD_SECT_REL_BSS,
+
+	MOD_SECT_SYMTAB,
+	MOD_SECT_STRTAB,
+	MOD_SECT_SHSTRTAB,
+
+	/* Should always be last */
+	MOD_SECT_COUNT,
+};
+
+/**
+ * @brief Loadable code module
+ */
+struct module {
+	sys_snode_t _mod_list;
+
+	/** Name of the module */
+	char name[16];
+
+	/** Lookup table of module memory regions */
+	void *mem[MOD_MEM_COUNT];
+
+	/** Total size of the module memory usage */
+	size_t mem_size;
+
+	/** Exported symbols from the module, may be used in other modules */
+	struct module_symtable sym_tab;
+};
+
+/**
+ * @brief Module loader context
+ *
+ * A source of an ELF stream/blob
+ */
+struct module_stream {
```

Comment:
In a future update to this the elf headers will be pointers and it will be entirely up to the stream implementation how that is worked with.

@lyakh is looking to load and keep the elf in memory and do direct mapping of sections and headers without a copy, that's likely something you are interested in as well it seems like